CommonWallet/Core/Classes/Modules/TransferConfirmation/TransferConfirmationPresenter.swift

```diff
@@ -57,6 +70,90 @@ final class TransferConfirmationPresenter {
                                                                                       locale: locale)
         view?.didReceive(accessoryViewModel: accessoryViewModel)
     }
+
+    private func setupBalanceDataProvider() {
+        let changesBlock = { [weak self] (changes: [DataProviderChange<[BalanceData]>]) -> Void in
+            if let change = changes.first {
+                switch change {
+                case .insert(let items), .update(let items):
+                    self?.balances = items
+                default:
+                    break
+                }
+            }
+        }
+
+        let failBlock: (Error) -> Void = { [weak self] (error: Error) in
+            self?.logger?.error(error.localizedDescription)
+        }
+
+        let options = DataProviderObserverOptions(alwaysNotifyOnRefresh: true)
+        balanceDataProvider.addObserver(
+            self,
+            deliverOn: .main,
+            executing: changesBlock,
+            failing: failBlock,
+            options: options
+        )
+    }
+
+    private func setupMetadataProvider() {
+        let changesBlock = { [weak self] (changes: [DataProviderChange<TransferMetaData>]) -> Void in
+            if let change = changes.first {
+                switch change {
+                case .insert(let item), .update(let item):
+                    self?.metadata = item
+                default:
+                    break
+                }
+            }
+        }
+
+        let failBlock: (Error) -> Void = { [weak self] (error: Error) in
+            self?.logger?.error(error.localizedDescription)
+        }
+
+        let options = DataProviderObserverOptions(alwaysNotifyOnRefresh: true)
+        metadataProvider.addObserver(
+            self,
+            deliverOn: .main,
+            executing: changesBlock,
+            failing: failBlock,
+            options: options
+        )
+    }
+
+    private func validateTransfer(completion: (Result<Void, Error>) -> Void) {
+        guard let balances = balances, let metadata = metadata else {
+            // balances & metadata had been validated at previous screen
```

Comment:
Aren't they always validated on the first screen?

Reply:
You're right, but the problem is that user balance could change before transfer confirmation. Scenario:
1. Open Transfer
2. Proceed to Confirmation
3. Perform transfer with another app (i.e. polkadot js)
4. Back to fearless: Tap on `Confirm transfer` -> fails, but fees were paid (

Reply:
Sure, i now this issue, but i don't quite understand this success completion

Reply:
Yep, mostly likely due to poor connection the values might be present but those are from the cache and there is no difference with the check from the previous screen. Data provider works in a way that it first returns data from the cache and then makes a request to the backend.